solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java

```diff
@@ -252,7 +252,6 @@ private ZkWriteCommand updateState(final ClusterState prevState, ZkNodeProps mes
       log.info("Failed to update state because the replica does not exist, {}", message);
       return ZkStateWriter.NO_OP;
     }
-    boolean persistCollectionState = collection != null && collection.isPerReplicaState();
```

Comment:
Why is this removed?

If we remove this, it will result in unnecessary updates to `state.json`

Reply:
@noblepaul from my understanding of the current release/8.8 [`ReplicaMutator`](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java#L255), it appears that:
1. For NON prs operation, this flag does not matter, it always return the ZkWriteCmd
2.For PRS operation, this flag is always true unless "collection is null" (which means someone is submitting `operation=state` message to OS Q when the collection is not yet exists - think it's not a valid case?)

Since in this current PR, we are correcting the logic of `ReplicaMutator#updateState` to ONLY perform `state.json` operation (no more PRS ops), both PRS and non PRS operation should return the same `ZkWriteCommand`, and such `persistCollectionState` flag does not have any reference anymore.

Please lemme know if any cases are overlooked 😄  . My assumption (which I will add javadoc) to this method now, is that it should only update state.json. If any initiator of such update (usually whoever submits the `operation=state` to OS Q) should be aware of exactly what it's trying to update. If it's PRS collection, it should determine if such update would affect PRS entries, state.json or both and do the corresponding actions accordingly (send `operation=state` to OS Q and/or update the PRS entries directly).

Reply:
> Since in this current PR, we are correcting the logic of ReplicaMutator#updateState to ONLY perform state.json operation (no more PRS ops), both PRS and non PRS operation should return the same ZkWriteCommand

That is not true. The state message can change the `state` flag which results in modification of `state.json` and that is unnecessary for PRS collections.

Reply:
@noblepaul @patsonluk  would be good to see unit test for it first.

Reply:
@hiteshk25 there is a [testcase](https://github.com/fullstorydev/lucene-solr/blob/3a8818ca1fb209156417ed18ccc3cdabc1edf260/solr/core/src/test/org/apache/solr/core/FSPRSTest.java#L333) already in this #168

Reply:
> > Since in this current PR, we are correcting the logic of ReplicaMutator#updateState to ONLY perform state.json operation (no more PRS ops), both PRS and non PRS operation should return the same ZkWriteCommand
> 
> That is not true. The state message can change the `state` flag which results in modification of `state.json` and that is unnecessary for PRS collections.

In the existing code as far as `collection.isPerReplicaState()` is `true`, then such flag can only be `true` (since all the other assignment assigns `true`), that means the existing code seems to always updates `state.json` for PRS collection, is that a behavior we want to change as a part of this fix?

Reply:
>hat means the existing code seems to always updates `state.json` for PRS collection, is that a behavior we want to change as a part of this fix?

Yes. absolutely. In a PRS collection, there is no point in updating the `state` flag of a replica in `state.json` as it is always overridden by the value in the individual replica states

Reply:
Thanks for the explanation @noblepaul ! 💡 My original intention of this PR is to focus on fixing the the race condition - enforcing PRS ops in data node and remove it from overseer replica mutator, and leave the rest of the behaviors exactly the same as before. 

I'm okay if we want to change this as a part of the PR too, that appears to be an optimization (fixing redundant calls as in the existing code?). what do u think @hiteshk25 😊

Reply:
Went into a bit more details this morning to assess the scope of the redundant state.json writes for our common usages in FS. I have found:
- Collection Creation
- Add replica (solrman move operation)

Since both of the above creates a new core which triggers `ZkController#publish` [with `forcePublish` as `true`](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1905), which indeed would issue a `state.json` update when replica state is set as `DOWN` . Such update is indeed redundant/unnecessary as @noblepaul pointed out.

Out of the above cases, the redundant update on `state.json` for collection creation could be more of a concern (as batching is not enabled in ZkStateWriter PRS collection), therefore creating a PRS collection with n shard could potentially issue n unnecessary writes to state.json?

The redundant state.json write however, does not affect:
- Solr service stop (in our case SIGTERM, if unsuccessful SIGKILL)
- Solr service start (`ZkController#preRegister` this [condition](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1904) is `false`, then on `ZkController#register` state `ACTIVE`, `forcePubish` and `sendToOverseer` are `false` at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1704) hence no message to OS)

Reply:
> herefore creating a PRS collection with n shard could potentially issue n unnecessary writes to state.json? 

true

> Solr service start (ZkController#preRegister this [condition](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1904) is false, then on ZkController#register state ACTIVE, forcePubish and sendToOverseer are false at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1704) hence no message to OS)

True. But for our JUnit tests, restart always start the node in another port and ends up changing the `node_name` & `base_url` attributes